### PR TITLE
R2 score metric

### PR DESCRIPTION
**Pull request by kafkdev**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/github/catalyst-team/catalyst/blob/master/examples/notebooks/colab_ci_cd.ipynb) with `latest` and `minimal` requirements?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [ ] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [ ] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [ ] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [ ] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

**Review by Scitator — COMMENTED** [review]
LGTM, nevertheless, we need to add a few things:
1. correct code style with `catalyst-make-codestyle`
2. tests :)
3. Metric API extension?
4. Callback API extension?

File: catalyst/metrics/functional/_r2_score.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            outputs=torch.tensor([2.5, 0.0, 2, 8]),
+            targets=torch.tensor([3, -0.5, 2, 7]),
+        )
+        # tensor([0.9486])
+    """
+    total_sum_of_squares = torch.sum(
+        torch.square(targets.float() - torch.mean(targets.float()))
+    ).view(-1)
+    residual_sum_of_squares = torch.sum(torch.square(targets.float() - outputs.float())).view(-1)
+    output = 1 - residual_sum_of_squares / total_sum_of_squares
```
Comment (Podidiving) [inline catalyst/metrics/functional/_r2_score.py]: I believe, you should do something like `max(total_sum_of_squares, eps)` to avoid zero division

File: catalyst/metrics/functional/_r2_score.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from typing import Sequence
+
+import torch
+
+
+def r2_score(outputs: torch.Tensor, targets: torch.Tensor,) -> Sequence[torch.Tensor]:
```
Comment (Podidiving) [inline catalyst/metrics/functional/_r2_score.py]: mb it should be done like sklearn implementation https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html ?
I mean, it seems cool to have possibility to calculate r2 score of tensors with shape [bs, num_outputs] right in torch
plus, `sample_weight` might be useful

File: tests/catalyst/metrics/functional/test_r2_squared.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+from catalyst.metrics.functional._r2_squared import r2_squared
+
+
+def test_r2_squared():
+    """
+  Tests for catalyst.metrics.r2_squared metric.
+  """
+    y_true = torch.tensor([2.5, 0.0, 2, 8])
+    y_pred = torch.tensor([3, -0.5, 2, 7])
+    val = r2_squared(y_true, y_pred)
```
Comment (Scitator) [inline tests/catalyst/metrics/functional/test_r2_squared.py]: you have `r2_squared(outputs: torch.Tensor, targets: torch.Tensor)`, rather than `targets, outputs`

File: catalyst/metrics/_r2_squared.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        self.y_sum += torch.sum(y_true)
+        self.y_sq_sum += torch.sum(torch.square(y_true))
+
+    def compute(self) -> torch.Tensor:
+        """
+        Return accumulated metric
+        """
+        return 1 - self.delta_sum / (self.y_sq_sum - (self.y_sum ** 2) / self.num_examples)
+
+    def compute_key_value(self) -> torch.Tensor:
```
Comment (Scitator) [inline catalyst/metrics/_r2_squared.py:55]: should be key-value here ;)

**Review by Scitator — DISMISSED** [review]
btw, could you please add all of them to the required `__init__`s?

File: tests/catalyst/metrics/functional/test_r2_squared.py (hunk 3 above)
Comment (kafkdev) [reply tests/catalyst/metrics/functional/test_r2_squared.py]: Whoops

**Review by Scitator — DISMISSED** [review]
